org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/GenericsRegressionTest.java

```diff
@@ -6681,5 +6682,86 @@ public void testBug576524() {
 		);
 	}
 }
+
+// https://github.com/eclipse-jdt/eclipse.jdt.core/issues/472
+public void testBugGH472_a() {
+	if (this.complianceLevel >= ClassFileConstants.JDK1_8) {
+		this.runConformTest(
+			new String[] {
+				"ReturnTypeTest.java",
+				"public class ReturnTypeTest {\n"
+				+ "    <T> T m(Class<T> arg1, Class<T> arg2) { return null; }\n"
+				+ "\n"
+				+ "    void test(Class c1, Class<Class<String>> c2) throws Exception {\n"
+				+ "        m(c1, c2).newInstance();\n"
+				+ "    }\n"
+				+ "}"
+			}
+		);
+	}
+}
+
+public void testBugGH472_b() {
+	if (this.complianceLevel >= ClassFileConstants.JDK1_8) {
+		this.runConformTest(
+			new String[] {
+				"ReturnTypeTest.java",
+				"public class ReturnTypeTest {\n"
+				+ "    <T> T m(Class<T> arg1, Class<? extends T> arg2) { return null; }\n"
+				+ "\n"
+				+ "    void test(Class c1, Class<Class<String>> c2) throws Exception {\n"
+				+ "        m(c1, c2).newInstance();\n"
+				+ "    }\n"
+				+ "}"
+			}
+		);
+	}
+}
+
+public void testBugGH472_c() {
+	if (this.complianceLevel >= ClassFileConstants.JDK1_8) {
+		this.runConformTest(
+			new String[] {
+				"ReturnTypeTest.java",
+				"public class ReturnTypeTest {\n"
+				+ "    <T> T m(Class<T> arg1, Class<? super T> arg2) { return null; }\n"
+				+ "\n"
+				+ "    void test(Class c1, Class<Class<String>> c2) throws Exception {\n"
+				+ "        m(c1, c2).newInstance();\n"
+				+ "    }\n"
+				+ "}"
+			}
+		);
+	}
+}
```

Comment:
For the above group of similar tests it would be lovely if a short comment would highlight the differences / focus per test case, to help future readers :smile:

Reply:
sure.